internal/backend/orgs/orgs.go

```diff
@@ -99,54 +99,96 @@ func (o *orgs) ListMembers(ctx context.Context, id sdktypes.OrgID) ([]sdktypes.U
+func (o *orgs) AddMember(ctx context.Context, oid sdktypes.OrgID, uid sdktypes.UserID, s sdktypes.OrgMemberStatus) error {
 	if !uid.IsValid() {
 		uid = authcontext.GetAuthnUserID(ctx)
 	}
 
-	if err := authz.CheckContext(ctx, oid, "write:add-user"); err != nil {
+	if s == sdktypes.OrgMemberStatusUnspecified {
+		s = sdktypes.OrgMemberStatusInvited
+	}
+
+	if err := authz.CheckContext(
+		ctx,
+		oid,
+		"write:add-member",
+		authz.WithAssociationWithID("user", uid),
+		authz.WithData("status", s.String()),
+	); err != nil {
 		return err
 	}
 
-	return AddMember(ctx, o.db, oid, uid)
+	return AddMember(ctx, o.db, oid, uid, s)
 }
 
 func (o *orgs) RemoveMember(ctx context.Context, oid sdktypes.OrgID, uid sdktypes.UserID) error {
-	if err := authz.CheckContext(ctx, oid, "write:rm-user"); err != nil {
+	if !uid.IsValid() {
+		uid = authcontext.GetAuthnUserID(ctx)
+	}
+
+	if err := authz.CheckContext(ctx, oid, "delete:remove-member", authz.WithAssociationWithID("user", uid)); err != nil {
 		return err
 	}
 
 	return o.db.RemoveOrgMember(ctx, oid, uid)
 }
 
-func (o *orgs) IsMember(ctx context.Context, oid sdktypes.OrgID, uid sdktypes.UserID) (bool, error) {
+func (o *orgs) GetMemberStatus(ctx context.Context, oid sdktypes.OrgID, uid sdktypes.UserID) (sdktypes.OrgMemberStatus, error) {
 	if !uid.IsValid() {
 		uid = authcontext.GetAuthnUserID(ctx)
 	}
 
-	if err := authz.CheckContext(ctx, oid, "read:is-member"); err != nil {
-		return false, err
-	}
-
-	if err := authz.CheckContext(ctx, uid, "read:is-org-member"); err != nil {
-		return false, err
+	if err := authz.CheckContext(
+		ctx,
+		uid,
+		"read:get-org-member-status",
+		authz.WithAssociationWithID("org", oid),
+		authz.WithConvertForbiddenToNotFound,
+	); err != nil {
+		return sdktypes.OrgMemberStatusUnspecified, err
 	}
 
-	return o.db.IsOrgMember(ctx, oid, uid)
+	return o.db.GetOrgMemberStatus(ctx, oid, uid)
 }
 
-func (o *orgs) GetOrgsForUser(ctx context.Context, uid sdktypes.UserID) ([]sdktypes.Org, error) {
+func (o *orgs) GetOrgsForUser(ctx context.Context, uid sdktypes.UserID) ([]*sdkservices.OrgWithMemberStatus, error) {
 	if !uid.IsValid() {
 		uid = authcontext.GetAuthnUserID(ctx)
 	}
 
-	if err := authz.CheckContext(ctx, uid, "read:get-orgs"); err != nil {
+	if err := authz.CheckContext(ctx, uid, "read:get-orgs", authz.WithConvertForbiddenToNotFound); err != nil {
 		return nil, err
 	}
 
 	return o.db.GetOrgsForUser(ctx, uid)
 }
+
+func (o *orgs) UpdateMemberStatus(ctx context.Context, oid sdktypes.OrgID, uid sdktypes.UserID, status sdktypes.OrgMemberStatus) error {
+	if !uid.IsValid() {
+		uid = authcontext.GetAuthnUserID(ctx)
+	}
+
+	return o.db.Transaction(ctx, func(tx db.DB) error {
+		curr, err := tx.GetOrgMemberStatus(ctx, oid, uid)
+		if err != nil {
+			return err
+		}
+
+		if err := authz.CheckContext(
+			ctx,
+			oid,
+			"write:update-member-status",
+			authz.WithAssociationWithID("user", uid),
+			authz.WithData("next_status", status.String()),
```

Comment:
"Next" implies a sequence of states, or an order, why not just "new_status"? If you change this, remember to change the Rego policy file too.

Reply:
will change to `new`

Reply:
done